Source/DafnyRuntime/DafnyRuntimeJava/src/main/java/dafny/UInt.java

```diff
@@ -0,0 +1,148 @@
+        return Integer.compareUnsigned(x.inner, y.inner);
+    }
+
+    public int compareTo(UInt other){
+        return Integer.compareUnsigned(inner, other.inner);
+    }
+
+    public int value(){
+        return inner;
+    }
+
+    public int intValue(){
+        return inner;
+    }
+
+    public double doubleValue(){
+        return (double) Integer.toUnsignedLong(inner);
+    }
+
+    public float floatValue(){
+        return (float) Integer.toUnsignedLong(inner);
+    }
+
+    public long longValue(){
+        return Integer.toUnsignedLong(inner);
+    }
+
+    //Invariant that other.inner is positive, so no underflow check needed
+    //todo: determine what to do if there is overflow ever
+    //addExact will throw an error if there is overflow
+    public UInt add(UInt other){
+        assert Integer.toUnsignedLong(inner) + Integer.toUnsignedLong(other.inner) <= 0xffffffffl : "Precondition Failure";
+        return new UInt(inner+other.inner);
+    }
+
+    //Invariant that other.inner is positive, so no overflow check needed
+    public UInt subtract(UInt other){
+        assert Integer.compareUnsigned(inner, other.inner) >= 0: "Precondition Failure";
+        return new UInt(inner-other.inner);
+    }
+
+    //Invariant that other.inner is positive, so no underflow check needed
+    public UInt multiply(UInt other){
+        assert Integer.toUnsignedLong(inner) * Integer.toUnsignedLong(other.inner) <= 0xffffffffl : "Precondition Failure";
+        return new UInt(inner*other.inner);
+    }
+
+    //Invariant that other.inner is positive, so only nonzero check needed
+    public UInt divide(UInt other){
+        assert other.inner != 0 : "Precondition Failure";
+        return new UInt(Integer.divideUnsigned(inner, other.inner));
+    }
+
+    //Invariant that other.inner is positive, so only nonzero check needed
+    public UInt mod(UInt other){
+        assert other.inner != 0 : "Precondition Failure";
+        return new UInt(Integer.remainderUnsigned(inner, other.inner));
+    }
+
+    @Override
+    public boolean equals(Object obj) {
+        if (this == obj) return true;
+        if (obj == null) return false;
+        if (getClass() != obj.getClass()) return false;
+        UInt o = (UInt) obj;
+        return inner == o.inner;
+    }
+
+    @Override
+    public int hashCode() {
+        return Integer.hashCode(inner);
+    }
+
+    @Override
+    public String toString() {
+        return Integer.toUnsignedString(inner);
+    }
+
+    public UInt xor(UInt other){
+        return new UInt(inner ^ other.inner);
+    }
+
+    public UInt or(UInt other){
+        return new UInt(inner | other.inner);
+    }
+
+    public UInt and(UInt other){
+        return new UInt(inner & other.inner);
+    }
+
+    public UInt not(){
+        return new UInt(~inner);
+    }
+
+    public UInt shiftLeft(int i){
+        return new UInt(inner << i);
+    }
+
+    public UInt shiftRight(int i){
+        return new UInt(inner >> i);
```

Comment:
This should be `>>>`. Please include the following tests (either in `Test/dafny0/Bitvectors.dfy` or in a new test file):

```
method Main() {
  var u: bv32 := 0xFFFF_FFFF;
  var v := u >> 1;
  Print([u, v]);  // 0xFFFF_FFFF 0x7FFF_FFFF
}

method Print(s: seq<bv32>) {
  var i := 0;
  while i < |s| {
    PrintAsHex(s[i] as uint);
    print " ";
    i := i + 1;
  }
  print "\n";
}

type uint = x | 0 <= x < 0x1_0000_0000

method PrintAsHex(u: uint) {
  var s, i, u := "", 0, u;
  while i < 8 {
    var d := u % 16;
    u := u / 16;
    var ch := if d < 10 then '0' + d as char else 'A' + (d - 10) as char;
    if i == 4 {
      s := "_" + s;
    }
    s := [ch] + s;
    i := i + 1;
  }
  print "0x", s;
}
```

The same comment applies to ULong.java, UShort.java, and UByte.java.